cli/src/main/scala/com/codacy/analysis/cli/formatter/Sarif.scala

```diff
@@ -0,0 +1,204 @@
+package com.codacy.analysis.cli.formatter
+
+import java.io.PrintStream
+import java.math.BigInteger
+import java.net.URI
+import java.security.MessageDigest
+import java.util
+
+import better.files.File
+import com.codacy.analysis.core.model.{Issue, Result}
+import com.codacy.plugins.api.results.Pattern
+import com.codacy.plugins.api.{PatternDescription, results}
+import io.circe.generic.auto._
+import io.circe.syntax._
+import io.circe.{Encoder, Printer}
+
+import scala.collection.JavaConverters._
+import scala.util.matching.Regex
+
+object Sarif extends FormatterCompanion {
+  override val name: String = "sarif"
+
+  override def apply(outputStream: PrintStream, executionDirectory: File): Formatter =
+    new Sarif(outputStream, executionDirectory)
+}
+
+private[formatter] class Sarif(val stream: PrintStream, val executionDirectory: File) extends Formatter {
+
+  val schema = new URI("https://docs.oasis-open.org/sarif/sarif/v2.1.0/cos02/schemas/sarif-schema-2.1.0.json")
+  val version = "2.1.0"
+
+  private val rootDirectory = "file:///codacy"
+
+  private val spacesCompiledRegex: Regex = "\\s".r
+
+  private implicit val levelEncoder: Encoder[SarifReport.Level.Value] = Encoder.encodeEnumeration(SarifReport.Level)
+
+  private val runs: util.ArrayList[SarifReport.Run] = new util.ArrayList()
+
+  override def begin(): Unit = ()
+
+  override def end(): Unit = {
+    val sarifReport = SarifReport(`$schema` = schema.toString, version = version, runs = runs.asScala.toList)
+    val printer = Printer.noSpaces.copy(dropNullValues = true)
+    val reportString = printer.print(sarifReport.asJson)
+    stream.print(reportString)
+    stream.flush()
+
+    ()
+  }
+
+  override def addAll(toolSpecification: Option[com.codacy.plugins.api.results.Tool.Specification],
+                      patternDescriptions: Set[PatternDescription],
+                      results: Seq[Result]): Unit = {
```

Comment:
This method is currently 80 lines long.. can we break some of these steps into private methods to improve readability?

Reply:
Done